src/Sonarr.Api.V5/EpisodeFiles/EpisodeFileController.cs

```diff
@@ -0,0 +1,202 @@
+using NzbDrone.Core.DecisionEngine.Specifications;
+using NzbDrone.Core.Exceptions;
+using NzbDrone.Core.Languages;
+using NzbDrone.Core.MediaFiles;
+using NzbDrone.Core.MediaFiles.Events;
+using NzbDrone.Core.Messaging.Events;
+using NzbDrone.Core.Parser;
+using NzbDrone.Core.Parser.Model;
+using NzbDrone.Core.Tv;
+using NzbDrone.SignalR;
+using Sonarr.Http;
+using Sonarr.Http.REST;
+using Sonarr.Http.REST.Attributes;
+using BadRequestException = Sonarr.Http.REST.BadRequestException;
+
+namespace Sonarr.Api.V5.EpisodeFiles;
+
+[V5ApiController]
+public class EpisodeFileController : RestControllerWithSignalR<EpisodeFileResource, EpisodeFile>,
+                             IHandle<EpisodeFileAddedEvent>,
+                             IHandle<EpisodeFileDeletedEvent>
+{
+    private readonly IMediaFileService _mediaFileService;
+    private readonly IDeleteMediaFiles _mediaFileDeletionService;
+    private readonly ISeriesService _seriesService;
+    private readonly ICustomFormatCalculationService _formatCalculator;
+    private readonly IUpgradableSpecification _upgradableSpecification;
+
+    public EpisodeFileController(IBroadcastSignalRMessage signalRBroadcaster,
+                         IMediaFileService mediaFileService,
+                         IDeleteMediaFiles mediaFileDeletionService,
+                         ISeriesService seriesService,
+                         ICustomFormatCalculationService formatCalculator,
+                         IUpgradableSpecification upgradableSpecification)
+        : base(signalRBroadcaster)
+    {
+        _mediaFileService = mediaFileService;
+        _mediaFileDeletionService = mediaFileDeletionService;
+        _seriesService = seriesService;
+        _formatCalculator = formatCalculator;
+        _upgradableSpecification = upgradableSpecification;
+    }
+
+    protected override EpisodeFileResource GetResourceById(int id)
+    {
+        var episodeFile = _mediaFileService.Get(id);
+        var series = _seriesService.GetSeries(episodeFile.SeriesId);
+
+        var resource = episodeFile.ToResource(series, _upgradableSpecification, _formatCalculator);
+
+        return resource;
+    }
+
+    [HttpGet]
+    [Produces("application/json")]
+    public List<EpisodeFileResource> GetEpisodeFiles(int? seriesId, [FromQuery] List<int>? episodeFileIds)
+    {
+        if (!seriesId.HasValue && episodeFileIds?.Any() == false)
+        {
+            throw new BadRequestException("seriesId or episodeFileIds must be provided");
+        }
+
+        if (seriesId.HasValue)
+        {
+            var series = _seriesService.GetSeries(seriesId.Value);
+            var files = _mediaFileService.GetFilesBySeries(seriesId.Value);
+
+            if (files == null)
+            {
+                return new List<EpisodeFileResource>();
+            }
+
+            return files.ConvertAll(e => e.ToResource(series, _upgradableSpecification, _formatCalculator))
+                        .ToList();
```

Comment:
Not needed since ConvertAll already returns a list.

```suggestion
            return files.ConvertAll(e => e.ToResource(series, _upgradableSpecification, _formatCalculator));
```